### PR TITLE
8281375: Accelerate bitCount operation for AVX2 and AVX512 target.

### PR DESCRIPTION
**Pull request by jatin-bhateja**
Summary of changes:

- Patch extends existing vectorized bitCount optimization added with [JDK-8278868](https://bugs.openjdk.java.net/browse/JDK-8278868) and emits optimized JIT sequence for AVX2 and other AVX512 targets which do not support avx512_vpopcntdq feature.
- Since PopCountVI/PopCountVL node emit different instruction sequence based on the target features hence a rudimentary cost mode has been added which influences the SLP unrolling factor to prevent generating bloated main loops.
- Following are the performance results of an existing [JMH micro](https://github.com/jatin-bhateja/jdk/blob/master/test/micro/org/openjdk/bench/vm/compiler/VectorBitCount.java) over various X86 targets.

Benchmark | SIZE | Baseline AVX2 (ns/op) | Withopt AVX2 (ns/op) | Gain % | Baseline AVX3 (ns/op) | Withopt AVX3 (ns/op) | Gain % | Baseline AVX3 (VPOPCOUNTDQ) | Withopt AVX3 (VPOCOUNTDQ) | Gain %
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
VectorBitCount.WithSuperword.intBitCount | 1024 | 1089.799 | 420.156 | 159.3796114 | 1083.92 | 203.958 | 431.442748 | 88.958 | 60.096 | 48.02649095
VectorBitCount.WithSuperword.longBitCount | 1024 | 417.458 | 413.859 | 0.869619846 | 417.203 | 214.949 | 94.09394787 | 105.954 | 117.019 | -9.455729411

Please review and share your feedback.

Best Regards,
Jatin

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8281375](https://bugs.openjdk.java.net/browse/JDK-8281375): Accelerate bitCount operation for AVX2 and AVX512 target.

### Reviewers
 * [Sandhya Viswanathan](https://openjdk.java.net/census#sviswanathan) (@sviswa7 - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7373/head:pull/7373` \
`$ git checkout pull/7373`

Update a local copy of the PR: \
`$ git checkout pull/7373` \
`$ git pull https://git.openjdk.java.net/jdk pull/7373/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7373`

View PR using the GUI difftool: \
`$ git pr show -t 7373`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7373.diff">https://git.openjdk.java.net/jdk/pull/7373.diff</a>

</details>

### CODE REVIEW

**Review by neliasso — COMMENTED** [review]
Hi Jatin, nice work!

One quick comment - the _vector_popcount_lut it's generated unconditionally - could you guard it so that it's only generated when it can be used? My preferred choice would to be have it be generated lazily, but that's an enhancement all of it's own.

Best regards,
Nils

File: src/hotspot/share/opto/loopTransform.cpp
Comment (theRealELiu) [inline src/hotspot/share/opto/loopTransform.cpp:976]: For other platforms, I think `vector_op_cost` should be better to return 0 at this moment.

**Comment by merykitty** [comment]
Hi,

I think you can dump a lane into the constant table and use `vpbroadcasti128` for the look-up table.

Thanks.

**Comment by jatin-bhateja** [comment]
> Hi,
> 
> I think you can dump a lane into the constant table and use `vpbroadcasti128` for the look-up table.
> 
> Thanks.

That will be generated per method and proposal is still not final in P

File: src/hotspot/cpu/x86/matcher_x86.hpp
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/matcher_x86.hpp:194]: Currently its a rough approximation of generated instruction size over X86 since its mainly used to influence SLP unrolling factor.

File: src/hotspot/share/opto/loopTransform.cpp
Comment (theRealELiu) [reply src/hotspot/share/opto/loopTransform.cpp:976]: I thought the total cost has been counted by `uint body_size = _body.size()`. So that for other platforms, the total cost now is one more if counted PopCountVI/VL.

File: src/hotspot/cpu/x86/matcher_x86.hpp
Comment (theRealELiu) [reply src/hotspot/cpu/x86/matcher_x86.hpp:194]: Thanks! My concern is that the name "vector_op_cost", which looks like very general but it only cares about size overhead at this moment. In other place maybe we need to concern the time overhead. I prefer to give the function a more specific name.

**Comment by merykitty** [comment]
We recently have [JDK-8278947](https://bugs.openjdk.java.net/browse/JDK-8278947) for this. With some adjustment you can ask for alignment of the constant, too.
Thanks.

**Comment by jatin-bhateja** [comment]
> We recently have [JDK-8278947](https://bugs.openjdk.java.net/browse/JDK-8278947) for this. With some adjustment you can ask for alignment of the constant, too. Thanks.

- Emitting 64 byte constants may prove to be costly, given that code-size of a native method used while taking in-lining decisions takes constant table size into consideration,  also X86 instructions are variable sized (1-15 bytes) hence rematerializing constants using couple of instructions may be efficient in terms of both code size and latency. Given that loading a 64 byte constant not aligned to 64 byte address boundary may  span across multiple cache lines.
- But we can evaluate and consider extensions done by your patch separately.

**Comment by merykitty** [comment]
Since the lookup table is the replicate of 16 bytes, it is possible to just emit a 16-byte lane and broadcast it to the whole vector. IIRC a broadcast has the same cost as a normal load. I think it is up to you to decide if a 16-byte constant is reasonable but it is surely much more manageable than a 64-byte constant.

**Comment by jatin-bhateja** [comment]
> Since the lookup table is the replicate of 16 bytes, it is possible to just emit a 16-byte lane and broadcast it to the whole vector. IIRC a broadcast has the same cost as a normal load. I think it is up to you to decide if a 16-byte constant is reasonable but it is surely much more manageable than a 64-byte constant.

Hi 

Yes, it varies from case to case, like replicating 8 byte in a vector may be cheaper than loading from constant table. If we are able to rematerialize the value without touching the memory its should be the preference.  Also adding bulky constant table may have side effects which may impact other optimization which considers native code size as a heuristic,  generating load time constant tables are sharable. We can make good use of enhancement in substituent patches.

Best Regards,
Jatin

**Comment by jatin-bhateja** [comment]
Hi @neliasso , your comments have been addressed.

File: src/hotspot/cpu/x86/stubGenerator_x86_32.cpp
Comment (neliasso) [inline src/hotspot/cpu/x86/stubGenerator_x86_32.cpp:4031]: Do we need it if VM_Version::supports_avx512_vpopcntdq()?

Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/stubGenerator_x86_32.cpp]: DONE

**Comment by jatin-bhateja** [comment]
Hi @neliasso , your comments have been addressed. Thanks

**Comment by jatin-bhateja** [comment]
Hi @neliasso, reviewers, just want to check if any other modification needed here. Regards, Jatin

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:4375]: This doesn't look correct for say 512 bit vector length. At this point, the xtmp2 has 64-bit popcount results for lower 8 integers and dst has 64-bit popcount results for upper 8 integers. The vpackuswb does interleaving between xtmp2 and dst at 128 bit lanes so the result is not in correct order.

Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:4375]: original vector of integer = [ a3  a2  a1  a0]
unpackldq =  [0, a1, 0, a0]
unpackhdq = [ 0, a3 , 0 , a2]
perform sum of absolute difference and store the result into LSB 16 bits of each quad word.
packuswb  packs at lane granularity i.e, 128 bits packed.
      128 bit               128 bit
[0, sa3, 0, sa2]  [ 0, sa1, 0, sa0 ]  =>   [ sa3, sa2 , sa1, sa0]

Comment (sviswa7) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:4375]: The problem is at 512 bit level.

Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:4375]:
```
original 512 bit vector holding 16 integers = [ a15 a14 a13 a12, a11 a10 a9 a8 ,a7 a6 a5 a4 ,a3 a2 a1 a0 ]
 unpackdq lower:
                512 bit (vec1) 
   128           128        128      128
 0 a13 0 a12  0 a9 0 a8  0 a5 0 a4  0 a1 0 a0
 unpackdq higher:
                512 bit (vec1) 
   128           128        128      128
 0 a15 0 a14  0 a11 0 a10  0 a7 0 a6  0 a3 0 a2
 Next sum of absolute difference operation followed by pack will squeez each 128 bit lane of two participant vectors
 and interleave them in resulatant vector.
         VEC1_L3     VEC1_L2     VEC1_L1   VEC1_L0              VEC2_L3     VEC2_L2      VEC2_L1    VEC2_L0           
 [ 0 sa13 0 sa12  0 sa9 0 sa8  0 sa5 0 sa4  0 sa1 0 sa0 ] [ 0 sa15 0 sa14  0 sa11 0 sa10  0 sa7 0 sa6  0 sa3 0 sa2 ]
 [ sa15 sa14 sa13 sa12   sa11 sa10 sa9 sa8  sa7 sa6 sa5 sa4  sa3 sa2 sa1 sa0]
```

**Comment by jatin-bhateja** [comment]
Hi @sviswa7 , your comments have been addressed.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:4360]: I think check for supports_avx512bw() is enough here. Do we need vl?

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86.ad:1870]: For vlen==16, check for only avx512bw() is needed.

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86.ad:1876]: In case of long vlen==8, check for only avx512bw() is needed.

**Comment by sviswa7** [comment]
@jatin-bhateja  The patch looks good to me.

**Review by sviswa7 — APPROVED** [review]

**Comment by PaulSandoz** [comment]
Running some tests.

**Comment by sviswa7** [comment]
Thanks a lot @PaulSandoz.

**Comment by PaulSandoz** [comment]
Tests pass.

**Comment by jatin-bhateja** [comment]
> Tests pass.

Thanks @PaulSandoz. X86 backend changes have been reviewed by @sviswa7, do we need second approval on this.

**Comment by PaulSandoz** [comment]
@jatin-bhateja yes, need another HS reviewer.

**Comment by sviswa7** [comment]
@neliasso Could you please review and approve the patch if it looks ok to you? Jatin has incorporated changes suggested by you earlier. Paul Sandoz has tested the patch.

**Comment by jatin-bhateja** [comment]
> @neliasso Could you please review and approve the patch if it looks ok to you? Jatin has incorporated changes suggested by you earlier. Paul Sandoz has tested the patch.

@sviswa7 , it seem @neliasso is on vacation.  @PaulSandoz  can anyone else review the patch.

**Comment by jatin-bhateja** [comment]
> > @neliasso Could you please review and approve the patch if it looks ok to you? Jatin has incorporated changes suggested by you earlier. Paul Sandoz has tested the patch.
> 
> @sviswa7 , it seem @neliasso is on vacation. @PaulSandoz can anyone else review the patch.

Hi @vnkozlov , can you kindly review this patch. Thanks

**Review by TobiHartmann — APPROVED** [review]
Looks good to me too.

**Comment by jatin-bhateja** [comment]
Thanks @TobiHartmann and @sviswa7.

**Pull request closed without merging**
